services/sidecar/rollup/derivation/batch_v0_encoder.go

```diff
@@ -0,0 +1,169 @@
+package derivation
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/specularL2/specular/services/sidecar/rollup/rpc/bridge"
+	spTypes "github.com/specularL2/specular/services/sidecar/rollup/types"
+	"github.com/specularL2/specular/services/sidecar/utils/fmt"
+	"github.com/specularL2/specular/services/sidecar/utils/log"
+)
+
+const v0 = 0x0
+
+var errBatchFull = errors.New("batch full")
+var errBatchTooSmall = errors.New("batch too small")
+
+type Config interface {
+	GetTargetBatchSize() uint64
+	GetSeqWindowSize() uint64
+	GetSubSafetyMargin() uint64
+}
+
+type BatchV0Encoder struct {
+	cfg      Config
+	timeout  uint64
+	currBuf  *bytes.Buffer
+	subBatch *subBatch
+}
+
+func NewBatchV0Encoder(cfg Config) *BatchV0Encoder {
+	return &BatchV0Encoder{cfg, 0, &bytes.Buffer{}, newSubBatch()}
+}
+
+func (e *BatchV0Encoder) GetBatch(l1Head spTypes.BlockID) ([]byte, error) {
+	if e.timeout != 0 && l1Head.GetNumber() >= e.timeout {
+		data := e.currBuf.Bytes()
+		e.currBuf.Reset()
+		return data, nil
+	}
+	if e.currBuf.Len() < int(e.cfg.GetTargetBatchSize()) {
+		return nil, errBatchTooSmall
+	}
+	data := e.currBuf.Bytes()
+	e.currBuf.Reset()
+	return data, nil
+}
+
+func (e *BatchV0Encoder) Reset() {
+	e.currBuf.Reset()
+	e.subBatch = newSubBatch()
+}
+
+// Processes a block. If the block is non-empty and fits, add it to the current sub-batch.
+// If the block would cause the batch to exceed the target size, close the entire batch (by writing to `currBuf“).
+// Returns an error if the block could not be processed.
+func (e *BatchV0Encoder) ProcessBlock(block *types.Block) error {
+	// Initialize buffer with version byte if it hasn't been already.
+	if e.currBuf.Len() == 0 {
+		if err := e.currBuf.WriteByte(e.getVersion()); err != nil {
+			return fmt.Errorf("failed to encode version: %w", err)
+		}
+	}
+	var (
+		// Block is empty
+		shouldSkipBlock = len(block.Transactions()) == 0
+		// Block would cause the batch to exceed the target size
+		shouldCloseBatch    = uint64(e.currBuf.Len())+e.subBatch.size() > e.cfg.GetTargetBatchSize()
+		shouldCloseSubBatch = shouldCloseBatch || (shouldSkipBlock && e.subBatch.size() > 0)
+	)
+	if shouldCloseSubBatch {
+		if err := e.closeSubBatch(); err != nil {
+			return fmt.Errorf("could not close sub-batch: %w", err)
+		}
+	}
+	// Enforce soft cap on batch size.
+	if shouldCloseBatch {
+		return errBatchFull
+	}
+	// Skip intrinsically-derivable blocks.
+	if shouldSkipBlock {
+		log.Info("Skipping intrinsically-derivable block", "block#", block.NumberU64())
+		return nil
+	}
+	// Decode oracle tx.
+	l1Epoch, _, _, _, _, err := bridge.UnpackL1OracleInput(block.Transactions()[0])
```

Comment:
this is wrong: block not always contain oracle transaction.
also, should you create a new subbatch when l1Epoch changes so that the oracle transaction is always at the first block in the subbatch